README.md
# StructuredReader

[![Gem Version](https://badge.fury.io/rb/structured_reader.svg)](https://badge.fury.io/rb/structured_reader)

This library allows you to create declarative rulesets (or schemas) for reading primitive data structures (hashes + arrays + strings + numbers) or JSON into validated data objects. Free yourself from `json.fetch(:widget).fetch(:box).fetch(:dimensions).fetch(:width)`. Get that good `widget.box.dimensions.width` without risking NoMethodErrors. Have confidence that if you're passed total unexpected nonsense, it won't be smoothed over by a convenient MagicHashyMash.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'structured_reader'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install structured_reader

## Usage

StructuredReader allows you to declare the expected structure of a JSON document (or any other data structure built on Ruby primitives). Readers are created with a simple domain-specific language (DSL) and are immutable. The declaration is evaluated at the time you create the reader, not every time it's used to read, so it's a good idea to assign it to a constant.

```ruby
require 'structured_reader'

READER = StructuredReader.json do |o|
  o.collection :widgets do |w|
    w.string :type, "widgetType"
    w.number :price
    w.string :description, nullable: true
    w.array :tags, of: :string
    w.object :dimensions, nullable: true do |dims|
      dims.number :weight
      dims.number :width
      dims.number :height
    end
    w.time :last_updated_at, "lastUpdated"
  end
  o.object :pagination, strict: true do |pg|
    pg.string :next_url, "nextUrl", nullable: true
    pg.number :total_items, "totalItems"
  end
end
```

Readers provide a `read` method, which takes either a Hash or a JSON string (which will be parsed and is expected to result in a Hash).

```ruby
document = {
  widgets: [
    {
      widgetType: "squorzit",
      price: 99.99,
      description: "who can even say?",
      tags: ["mysterious", "magical"],
      dimensions: {
        weight: 10,
        width: 5,
        height: 9001
      },
      lastUpdated: "2017-12-24 01:01 AM PST"
    },
    {
      widgetType: "frobulator",
      price: 0.79,
      tags: [],
      comment: "a bonus text",
      lastUpdated: "2017-12-24 01:05 AM PST"
    }
  ],
  pagination: {
    nextUrl: nil,
    totalItems: 2
  }
}

result = READER.read(document)
```

Hashes (objects) are parsed into Ruby Structs. Fields present in the object but not in the declaration are ignored. You can change this behavior by passing `strict: true` to `object`.

```ruby
p result.widgets.length # ==> 2
p result.widgets[0].dimensions.height # ==> 9001
p result.widgets[0].tags # ==> ["mysterious", "magical"]
p result.widgets[1].last_updated_at # ==> #<DateTime: 2017-12-24T01:05:00-08:00 ((2458112j,32700s,0n),-28800s,2299161j)>
p result.widgets[1].comment # ! ==> NoMethodError
```

## Validation

Readers validate input to ensure it is of the expected type:

```ruby
reader = StructuredReader.json do |o|
  o.object :point do |pt|
    pt.number :x
    pt.number :y
  end
end

reader.read({}) # ! ==> StructuredReader::WrongTypeError: expected a Hash, got a NilClass (at .point)
reader.read({point: {} }) # ! ==> StructuredReader::WrongTypeError: expected a Numeric, got a NilClass (at .point.x)
reader.read({point: { x: 1, y: "f" } }) # ! ==> StructuredReader::WrongTypeError: expected a Numeric, got a String (at .point.y)
```

A full validation mode is also available, which avoids exceptions:

```ruby
r = reader.validate({point: { x: nil, y: "f" } })
p r.ok? # ==> false
p r.errors # ==> [[".point.x", "expected a Numeric"], [".point.y", "expected a Numeric"]

r2 = reader.validate({point: { x: 5, y: 10 }})
p r2.object.point.x # ==> 5
```

## Reader Types

* `string`: The classic.
* `number`: Like in JSON, this can be an integer or a float.
* `object`: Must define at least one field.
* `array`: An array containing elements of a single type. Use `one_of` to support arrays of mixed types.
* `collection`: Shorthand for an `array` of `objects`.
* `one_of`: Takes several other reader types and tests them in order, returning the first one that succeeds.
* `literal`: Validates an exact, literal value (e.g. the field must contain "foo" and nothing but "foo"). Can be used with `one_of` to implement discriminated unions; see below.
* `null`: Shorthand for `literal nil`.
* `time`: Expects a String and parses it using DateTime.parse. I am currently thinking about a clean way to declare subtypes of string along the lines of "parsable with a provided parser".
* `raw`: Always validates and returns the unmodified value. May be useful if you have a "payload" or "attributes" structure that you want to pass along to somewhere else.
* `custom`: Define your own personal pan parser.

Reead the specs for more details.

## Advanced Declarations

### Discriminated Unions

It is common to have JSON data structures that use a type field to distinguish between objects with different fields:

```ruby
document = [
  {
    type: "square",
    length: 10,
  },
  {
    type: "rectangle",
    width: 5,
    height: 10,
  },
  {
    type: "circle",
    diameter: 4,
  },
]
```

This can be typed as a _discriminated union_ (or _tagged union_). The `one_of` and `literal` methods can be combined to validate and read this data. (A shorthand method for doing this may be implemented in the future.)

```ruby
reader = StructuredReader.json(root: :array) do |a|
  a.one_of do |shape|
    shape.object do |sq|
      sq.literal :type, value: "square"
      sq.number :length
    end
    shape.object do |rc|
      rc.literal :type, value: "rectangle"
      rc.number :width
      rc.number :height
    end
    shape.object do |cr|
      cr.literal :type, value: "circle"
      cr.number :diameter
    end
  end
end
```

Each element of the array is tested against each option, and the one that succeeds is used. If none match, `StructuredReader::WrongTypeError` will be raised. (If you need a fallback, consider the `raw` type.)

There are some performance implications here. In general, declarations are tested against the data in order. In this case, each element will be tested first to see if it is a square, then to see if it is a rectangle, then to see if it is a circle. It's a good idea when using `one_of` to ensure that the type field (or some other distinguishing field) is declared first, so the tests can fail as fast as possible.

```ruby
result = reader.read(document)

p result.length # ==> 3
p result[0].type # ==> "square"
p result[0].length # ==> 10
p result[2].diameter # ==> 4
p result[0].diameter # ! ==> NoMethodError
```

## Defining Custom Types

The API for this is somewhat experimental. The best reference at this time is the specs.

## Caveats/Missing Features

### Stack depth

The depth of data structures you can read is limited by stack space, because the reading is done by mutual recursion. If you really have to parse a structure too deep for the stack, you can combine multiple layers of readers by using the `raw` type to pluck subdocuments out in unmodified form and then use a second reader to parse that.

### Performance

Unbenchmarked, unoptimized. It should be OK, though.

## Out Of Scope/Unfeatures

* Complex validation support. You can do a lot with `custom` fields, but I advise not pushing business logic into the serialization layer.
* Exact feature compatibility with JSON Schema. Same reason, really. You can use this library to validate JSON, but that's not its purpose.
* Inheritance/subtyping. Too much complexity for not much practical use. I might decide otherwise if I keep running into places it'd be useful.
* Tuples/fixed-length arrays. I don't see these in the wild very frequently, and I'm not a fan. `array`/`one_of` should suffice, and `custom` is there if you need it.

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/mboeh/structured_reader.

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGES.md
## 1.1 (2021-11-01)

* Support Ruby 3.x

## 1.0.1 (2017-12-26)

* Add a full validation reading mode that reports every detected error with the document.

## 1.0.0 (2017-12-24)

* First release!


spec/spec_helper.rb
require "bundler/setup"
require "structured_reader"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/structured_reader_spec.rb
require "spec_helper"

RSpec.describe StructuredReader do
  it "has a version number" do
    expect(StructuredReader::VERSION).not_to be nil
  end

  context ".json" do

    def reader(**args, &blk)
      StructuredReader.json(**args, &blk)
    end

    it "raises an exception if the declaration is empty" do
      object = {}
      expect{
        reader{}.read(object)
      }.to raise_error(StructuredReader::DeclarationError)
    end

    context "reading an object" do

      it "can read a nil" do
        object = { nul: nil }
        result = reader do |o|
          o.null :nul
        end.read(object)

        expect(result.nul).to be nil
      end

      it "can read a nullable field" do
        object = { str: nil }
        result = reader do |o|
          o.string :str, nullable: true
        end.read(object)

        expect(result.str).to be nil
      end

      it "can read a string" do
        object = { str: "bar" }
        result = reader do |o|
          o.string :str
        end.read(object)

        expect(result.str).to eq("bar")
      end

      it "can read an integer" do
        object = { num: 1 }
        result = reader do |o|
          o.number :num
        end.read(object)

        expect(result.num).to eq(1)
      end

      it "can read a float" do
        object = { num: 1.0 }
        result = reader do |o|
          o.number :num
        end.read(object)

        expect(result.num).to eq(1.0)
      end

      it "can read an array" do
        object = { ary: ["foo", "bar", "baz"] }
        result = reader do |o|
          o.array :ary do |a|
            a.string
          end
        end.read(object)

        expect(result.ary).to eq(["foo", "bar", "baz"])
      end

      it "can read a collection" do
        object = { coll: [{ foo: "bar" }] }
        result = reader do |o|
          o.collection :coll do |a|
            a.string :foo
          end
        end.read(object)

        expect(result.coll[0].foo).to eq("bar")
      end

      it "can read a nested object" do
        object = { obj: { foo: "bar" } }
        result = reader do |o|
          o.object :obj do |o2|
            o2.string :foo
          end
        end.read(object)

        expect(result.obj.foo).to eq("bar")
      end

      it "can enforce a strict set of fields" do
        rdr = reader do |o|
          o.object :obj, strict: true do |obj|
            obj.string :foo
          end
        end

        result = rdr.read({obj: {foo: "bar"}})

        expect(result.obj.foo).to eq("bar")

        expect{
          rdr.read({obj: {foo: "bar", baz: "bat"}})
        }.to raise_error(StructuredReader::WrongTypeError)
      end

    end

    context "reading an array" do

      it "can read an array of strings" do
        object = { ary: ["foo", "bar", "baz"] }
        result = reader do |o|
          o.array :ary, of: :string
        end.read(object)

        expect(result.ary).to eq(["foo", "bar", "baz"])
      end

      it "can read an array of objects" do
        object = { ary: [{ foo: "bar" }] }
        result = reader do |o|
          o.array :ary do |a|
            a.object do |o2|
              o2.string :foo
            end
          end
        end.read(object)

        expect(result.ary[0].foo).to eq("bar")
      end

    end

    context "reading a raw value" do

      it "returns the literal, unmodified value" do
        object = { raw: [ { yep: 1 }, { what_of_it: ["woot"] }, true ] }
        result = reader do |o|
          o.raw :raw
        end.read(object)

        expect(result.raw).to eq(object[:raw])
      end

    end

    context "reading a literal value" do

      it "returns the literal, unmodified value if it matches the provided value" do
        object = { lit: "yes" }
        result = reader do |o|
          o.literal :lit, value: "yes"
        end.read(object)

        expect(result.lit).to eq("yes")
      end

      it "rejects any other value" do
        object = { lit: "no" }

        expect{
          reader do |o|
            o.literal :lit, value: "yes"
          end.read(object)
        }.to raise_error(StructuredReader::WrongTypeError)
      end

    end

    context "reading a custom value" do

      it "delegates the reading to a provided block" do
        rdr = reader do |o|
          o.custom :cust do |fragment, context|
            if fragment.kind_of?(Numeric) && (1..10).member?(fragment)
              context.accept(fragment)
            else
              context.flunk(fragment, "must be a number from 1 to 10")
            end
          end
        end

        result = rdr.read({ cust: 7 })
        expect(result.cust).to eq(7)

        expect{
          rdr.read({ cust: 11 })
        }.to raise_error(StructuredReader::WrongTypeError)
        expect{
          rdr.read({ cust: "7" })
        }.to raise_error(StructuredReader::WrongTypeError)
      end

    end

    context "reading any_of" do

      it "raises an exception if no options are provided" do
        expect{
          reader do |o|

          end.read({})
        }.to raise_error(StructuredReader::DeclarationError)
      end

      it "accepts any of the declared options" do
        rdr = reader do |o|
          o.one_of :vary do |v|
            v.string
            v.number
            v.array of: :string
          end
        end

        expect(
          rdr.read({vary: "foo"}).vary
        ).to eq("foo")
        expect(
          rdr.read({vary: 1}).vary
        ).to eq(1)
        expect(
          rdr.read({vary: ["hi", "there"]}).vary
        ).to eq(["hi", "there"])
      end

      it "rejects undeclared options" do
        rdr = reader do |o|
          o.one_of :vary do |v|
            v.string
          end
        end

        expect{
          rdr.read({vary: 1})
        }.to raise_error(StructuredReader::WrongTypeError)
      end

    end

    context "reading a discriminated union" do

      it "works" do
        object = {
          shapes: [
            {
              type: "square",
              length: 10,
            },
            {
              type: "rectangle",
              width: 5,
              height: 10,
            },
            {
              type: "circle",
              diameter: 4,
            },
          ]
        }
        rdr = reader do |o|
          o.array :shapes do |a|
            a.one_of do |shape|
              shape.object do |sq|
                sq.literal :type, value: "square"
                sq.number :length
              end
              shape.object do |rc|
                rc.literal :type, value: "rectangle"
                rc.number :width
                rc.number :height
              end
              shape.object do |cr|
                cr.literal :type, value: "circle"
                cr.number :diameter
              end
            end
          end
        end

        result = rdr.read(object)

        expect(result.shapes.length).to eq(3)
        expect(result.shapes[0].type).to eq("square")
        expect(result.shapes[0].length).to eq(10)
        expect(result.shapes[1].width).to eq(5)
        expect(result.shapes[2].diameter).to eq(4)
      end

    end

    context "defining new reader types" do

      it "can define a custom reader" do
        reader_set = StructuredReader.reader_set do |r|
          r.custom :score do |fragment, context|
            if fragment.kind_of?(Numeric) && (1..10).member?(fragment)
              context.accept(fragment)
            else
              context.flunk(fragment, "must be a number from 1 to 10")
            end
          end
        end

        rdr = reader(reader_set: reader_set) do |o|
          o.score :rank
        end

        result = rdr.read({ rank: 9 })
        expect(result.rank).to eq(9)
      end

      it "can define an object alias" do
        reader_set = StructuredReader.reader_set do |r|
          r.object :package_dimensions do |o|
            o.number :width
            o.number :height
            o.number :depth
            o.number :weight
          end
        end

        rdr = reader(reader_set: reader_set) do |o|
          o.package_dimensions :dims
        end

        result = rdr.read({
          dims: {
            width: 5,
            height: 4,
            depth: 4,
            weight: 20,
          }
        })
        expect(result.dims.weight).to eq(20)
      end

    end

    context ".validate" do

      it "returns a result object" do
        rdr = reader do |o|
          o.string :first_name
          o.string :last_name
        end

        result = rdr.validate({
          first_name: "Stanley",
          last_name: "Burrell"
        })

        expect(result.errors).to be_empty
        expect(result.object.first_name).to eq("Stanley")
        expect(result).to be_ok
      end

      it "returns a full list of errors" do
        rdr = reader do |o|
          o.string :first_name
          o.string :last_name
        end

        result = rdr.validate({
        })

        expect(result.object).to be_nil
        expect(result).not_to be_ok
        expect(result.errors).to eq([
          [".first_name", "expected a String"],
          [".last_name", "expected a String"]
        ])
      end

    end

  end

end


examples/selection.rb
require 'bundler/setup'

require 'structured_reader'

reader = StructuredReader.json do |o|
  o.collection :widgets do |w|
    w.string :type, "widgetType"
    w.number :price
    w.string :description, nullable: true
    w.array :tags, of: :string
    w.object :dimensions, nullable: true do |dims|
      dims.number :weight
      dims.number :width
      dims.number :height
    end
    w.time :last_updated_at, "lastUpdated"
  end
  o.object :pagination do |pg|
    pg.string :next_url, "nextUrl", nullable: true
    pg.number :total_items, "totalItems"
  end
end

document = {
  widgets: [
    {
      widgetType: "squorzit",
      price: 99.99,
      description: "who can even say?",
      tags: ["mysterious", "magical"],
      dimensions: {
        weight: 10,
        width: 5,
        height: 9001
      },
      lastUpdated: "2017-12-24 01:01 AM PST"
    },
    {
      widgetType: "frobulator",
      price: 0.79,
      tags: [],
      lastUpdated: "2017-12-24 01:05 AM PST"
    }
  ],
  pagination: {
    nextUrl: nil,
    totalItems: 2
  }
}

# This uses filename globbing for now, not proper XPath-style wildcards
context = StructuredReader::JSONReader::SelectionContext.new(".widgets\\[*\\].price")

result = reader.read(document, context)

p result


examples/exceptions.rb
require 'bundler/setup'

require 'structured_reader'

reader = StructuredReader.json do |o|
  o.collection :widgets do |w|
    w.string :type, "widgetType"
    w.number :price
    w.string :description, nullable: true
    w.array :tags, of: :string
    w.object :dimensions, nullable: true do |dims|
      dims.number :weight
      dims.number :width
      dims.number :height
    end
    w.time :last_updated_at, "lastUpdated"
  end
  o.object :pagination do |pg|
    pg.string :next_url, "nextUrl", nullable: true
    pg.number :total_items, "totalItems"
  end
end

document = {
  widgets: [
    {
      widgetType: "squorzit",
      price: 99.99,
      description: "who can even say?",
      tags: ["mysterious", "magical"],
      dimensions: {
        weight: 10,
        width: 5,
        height: nil
      },
      lastUpdated: "2017-12-24 01:01 AM PST"
    },
    {
      widgetType: "frobulator",
      price: "0.79",
      tags: [123, { foo: "bar" }],
      lastUpdated: nil
    }
  ],
  pagination: {
    nextUrl: nil,
    totalItems: 2
  }
}

result = reader.validate(document)

p result.ok?
p result.errors


lib/structured_reader.rb
# frozen_string_literal: true

require "structured_reader/version"
require "json"
require "date"

module StructuredReader
  Error = Class.new(StandardError)
  WrongTypeError = Class.new(Error)
  DeclarationError = Class.new(Error)

  def self.json(**kwargs, &blk)
    JSONReader.new(**kwargs, &blk)
  end

  def self.reader_set(&blk)
    JSONReader::ReaderSet.new.tap(&blk)
  end

  class JSONReader

    def initialize(root: :object, reader_set: ReaderSet.new, &blk)
      @root_reader = reader_set.reader(root, &blk)
    end

    def read(document, context = Context.new)
      if document.kind_of?(String)
        document = JSON.parse document
      end
      context.open do
        @root_reader.read(document, context)
      end
    end

    def validate(document)
      read(document, ValidatorContext.new)
    end

    class ObjectReader

      class ReaderBuilder

        def initialize(base, reader_set:)
          @base = base
          @reader_set = reader_set
        end

        def method_missing(type, name, field_name = name.to_s, *args, **kwargs, &blk)
          if @reader_set.has_reader?(type)
            @base.field name, field_name, @reader_set.reader(type, *args, **kwargs, &blk)
          else
            super
          end
        end

        def respond_to_missing?(type)
          @reader_set.has_reader?(type) || super
        end

      end

      def initialize(strict: false, reader_set:)
        @readers = {}
        @strict = strict
        yield ReaderBuilder.new(self, reader_set: reader_set)
        if @readers.empty?
          raise DeclarationError, "must define at least one field to read"
        end
        @object_klass = Struct.new(*@readers.keys)
        freeze
      end

      def read(fragment, context)
        if fragment.kind_of?(Hash)
          result = @object_klass.new
          @readers.each do |key, (field, reader)|
            value = fragment[field] || fragment[field.to_sym]
            context.push(".#{field}") do |sub_context|
              result[key] = reader.read(value, sub_context)
            end
          end
          if @strict && ((excess_keys = fragment.keys.map(&:to_sym) - @readers.keys)).any?
            return context.flunk(fragment, "found strictly forbidden keys #{excess_keys.inspect}")
          end
          result.freeze

          context.accept(result)
        else
          return context.flunk(fragment, "expected a Hash")
        end
      end

      def field(key, field_name, reader)
        @readers[key.to_sym] = [field_name, reader]
      end

    end

    class ArrayReader

      class ReaderBuilder

        def initialize(base, reader_set:)
          @base = base
          @reader_set = reader_set
        end

        def method_missing(type, *args, **kwargs, &blk)
          if @reader_set.has_reader?(type)
            @base.member @reader_set.reader(type, *args, **kwargs, &blk)
          else
            super
          end
        end

        def respond_to_missing?(type)
          @reader_set.has_reader?(type) || super
        end

      end

      def initialize(of: nil, reader_set:, &blk)
        if block_given?
          yield ReaderBuilder.new(self, reader_set: reader_set)
        elsif of
          ReaderBuilder.new(self, reader_set: reader_set).send(of)
        end

        unless @member_reader
          raise DeclarationError, "array must have a member type"
        end
      end

      def member(reader)
        @member_reader = reader
      end

      def read(fragment, context)
        if fragment.kind_of?(Array)
          context.accept(fragment.map.with_index do |member, idx|
            context.push("[#{idx}]") do |sub_context|
              @member_reader.read(member, sub_context)
            end
          end)
        else
          context.flunk(fragment, "expected an Array")
        end
      end

    end

    class CollectionReader < ArrayReader

      def initialize(**args, &blk)
        super do |a|
          a.object(&blk)
        end
      end

    end

    class NumberReader

      def initialize(**_)

      end

      def read(fragment, context)
        if fragment.kind_of?(Numeric)
          context.accept fragment
        else
          context.flunk(fragment, "expected a Numeric")
        end
      end

    end

    class StringReader

      def initialize(**_)

      end

      def read(fragment, context)
        if fragment.kind_of?(String)
          context.accept maybe_parse(fragment, context)
        else
          context.flunk(fragment, "expected a String")
        end
      end

      def maybe_parse(fragment, _context)
        fragment
      end

    end

    class TimeReader < StringReader

      def maybe_parse(fragment, context)
        begin
          context.accept DateTime.parse(fragment)
        rescue ArgumentError
          context.flunk(fragment, "could not be converted to a DateTime")
        end
      end

    end

    class LiteralReader

      def initialize(value:, **_)
        @value = value
      end

      def read(fragment, context)
        if fragment == @value
          context.accept fragment
        else
          context.flunk(fragment, "expected #{@value.inspect}")
        end
      end

    end

    class NullReader < LiteralReader

      def initialize(**_)
        super value: nil
      end

    end

    class RawReader

      def initialize(**_)

      end

      def read(fragment, context)
        context.accept fragment
      end

    end

    class CustomReader

      def initialize(**_, &blk)
        @read_action = blk
      end

      def read(fragment, context)
        @read_action.call(fragment, context)
      end

    end

    class OneOfReader

      class ReaderBuilder

        def initialize(base, reader_set:)
          @base = base
          @reader_set = reader_set
        end

        def method_missing(type, *args, **kwargs, &blk)
          if @reader_set.has_reader?(type)
            @base.option @reader_set.reader(type, *args, **kwargs, &blk)
          else
            super
          end
        end

        def respond_to_missing?(type)
          @reader_set.has_reader?(type) || super
        end

      end

      def initialize(reader_set:, **_)
        @readers = []
        yield ReaderBuilder.new(self, reader_set: reader_set)
        if @readers.empty?
          raise DeclarationError, "must define at least one option"
        end
        freeze
      end

      def read(fragment, context)
        @readers.each do |reader|
          result = ValidatorContext.new.open do |context|
            reader.read(fragment, context)
          end
          if result.ok?
            return context.accept(result.object)
          end
        end

        context.flunk(fragment, "was not any of the expected options")
      end

      def option(reader)
        @readers << reader
      end

    end

    class BuilderDeriver

      def initialize(klass, &blk)
        @klass = klass
        @build_action = blk
      end

      def new(*args, **kwargs)
        @klass.new(*args, **kwargs, &@build_action)
      end

    end

    class Context

      def initialize(where = "")
        @where = where.dup.freeze
      end

      def accept(fragment)
        fragment
      end

      def open(&blk)
        blk.call(self)
      end

      def flunk(fragment, reason)
        raise WrongTypeError, "#{reason}, got a #{fragment.class} (at #{@where})"
      end

      def push(path, &blk)
        yield self.class.new(@where + path)
      end

    end

    class ValidatorContext
      Result = Struct.new(:object, :errors) do
        def ok?
          errors.empty?
        end
      end

      def initialize(where = "", errors = [])
        @where = where.dup.freeze
        @errors = errors
      end

      def open(&blk)
        result = blk.call(self)
        Result.new(@errors.any? ? nil : result, @errors)
      end

      def accept(fragment)
        fragment
      end

      def flunk(fragment, reason)
        @errors << [@where, reason]
      end

      def push(path)
        yield self.class.new(@where + path, @errors)
      end

    end

    class SelectionContext

      def initialize(target, where = "", found = [])
        @target = target
        @where = where
        @found = found
      end

      def accept(fragment)
        if File.fnmatch(@target, @where)
          @found << fragment
          fragment
        else
          if @found.any?
            @found.first
          else
            unless @where.empty?
              fragment
            end
          end
        end
      end

      def flunk(fragment, reason)
        nil
      end

      def push(path)
        if @found.empty?
          yield self.class.new(@target, @where + path, @found)
        end
      end

    end

    class ReaderSet
      READERS = {
        array: ArrayReader,
        collection: CollectionReader,
        string: StringReader,
        time: TimeReader,
        object: ObjectReader,
        number: NumberReader,
        one_of: OneOfReader,
        null: NullReader,
        raw: RawReader,
        literal: LiteralReader,
        custom: CustomReader,
      }

      def initialize
        @readers = READERS.dup
      end

      def add_reader(type, reader)
        @readers[type.to_sym] = reader
      end

      def custom(type, &blk)
        add_reader type, BuilderDeriver.new(CustomReader, &blk)
      end

      def object(type, &blk)
        add_reader type, BuilderDeriver.new(ObjectReader, &blk)
      end

      def reader(type, *args, **kwargs, &blk)
        if kwargs[:nullable]
          kwargs = kwargs.dup
          kwargs.delete :nullable
          OneOfReader.new(reader_set: self) do |o|
            o.null
            o.send(type, *args, **kwargs, &blk)
          end
        else
          @readers.fetch(type).new(*args, reader_set: self, **kwargs, &blk)
        end
      end

      def has_reader?(type)
        @readers.has_key?(type)
      end
    end

  end

end


examples/example.rb
require 'bundler/setup'

require 'structured_reader'

reader = StructuredReader.json do |o|
  o.collection :widgets do |w|
    w.string :type, "widgetType"
    w.number :price
    w.string :description, nullable: true
    w.array :tags, of: :string
    w.object :dimensions, nullable: true do |dims|
      dims.number :weight
      dims.number :width
      dims.number :height
    end
    w.time :last_updated_at, "lastUpdated"
  end
  o.object :pagination do |pg|
    pg.string :next_url, "nextUrl", nullable: true
    pg.number :total_items, "totalItems"
  end
end

document = {
  widgets: [
    {
      widgetType: "squorzit",
      price: 99.99,
      description: "who can even say?",
      tags: ["mysterious", "magical"],
      dimensions: {
        weight: 10,
        width: 5,
        height: 9001
      },
      lastUpdated: "2017-12-24 01:01 AM PST"
    },
    {
      widgetType: "frobulator",
      price: 0.79,
      tags: [],
      lastUpdated: "2017-12-24 01:05 AM PST"
    }
  ],
  pagination: {
    nextUrl: nil,
    totalItems: 2
  }
}

result = reader.read(document)

p result.widgets.length # ==> 2
p result.widgets[0].dimensions.height # ==> 9001
p result.widgets[0].tags # ==> ["mysterious", "magical"]
p result.widgets[1].last_updated_at # ==> #<DateTime: 2017-12-24T01:05:00-08:00 ((2458112j,32700s,0n),-28800s,2299161j)>
